pymetkit/src/metkit/mars.py

```diff
@@ -0,0 +1,311 @@
+import os
+from cffi import FFI
+from pkg_resources import parse_version
+import findlibs
+from typing import IO, Iterator
+
+__metkit_version__ = "1.11.0"
+
+ffi = FFI()
+
+
+def ffi_encode(data) -> bytes:
+    if isinstance(data, bytes):
+        return data
+
+    if not isinstance(data, str):
+        data = str(data)
+
+    return data.encode(encoding="utf-8", errors="surrogateescape")
+
+
+def ffi_decode(data: FFI.CData) -> str:
+    buf = ffi.string(data)
+    if isinstance(buf, str):
+        return buf
+    else:
+        return buf.decode(encoding="utf-8", errors="surrogateescape")
+
+
+class Request:
+    def __init__(self, verb: str | None = None, **kwargs):
+        """
+        Create MetKit MarsRequest object. Parameters and values in
+        the request can be specified through kwargs, noting that
+        reserved words in Python must be suffixed with "_" e.g. "class_"
+        """
+        crequest = ffi.new("metkit_request_t **")
+        lib.metkit_new_request(crequest)
+        self.__request = ffi.gc(crequest[0], lib.metkit_free_request)
+        if verb is not None:
+            lib.metkit_request_set_verb(self.__request, ffi_encode(verb))
+        for param, values in kwargs.items():
+            self[param.rstrip("_")] = values
+
+    def ctype(self) -> FFI.CData:
+        return self.__request
+
+    def verb(self) -> str:
+        cverb = ffi.new("const char **")
+        lib.metkit_request_verb(self.__request, cverb)
+        return ffi_decode(cverb[0])
+
+    def expand(self, inherit: bool = True, strict: bool = False) -> "Request":
+        """
+        Return expanded request
+
+        Params
+        ------
+        inherit: bool, if True, populates expanded request with default values
+        strict: bool, if True, raise error instead of warning for invalid values
+
+        Returns
+        -------
+        Request, resulting from expansion
+        """
+        expanded_request = Request()
+        lib.metkit_request_expand(
+            self.__request, expanded_request.ctype(), inherit, strict
+        )
+        return expanded_request
+
+    def validate(self):
+        """
+        Check if request is valid against MARS language definition. Does not
+        inherit missing parameters.
+
+        Raises
+        ------
+        Exception if request is incompatible with MARS language definition
+        """
+        self.expand(False, True)
+
+    def keys(self) -> Iterator[str]:
+        """
+        Get iterator over parameters in request
+
+        Returns
+        -------
+        Iterator over parameter names
+        """
+        cparams = ffi.new("metkit_paramiterator_t **")
+        lib.metkit_request_params(self.__request, cparams)
+        self._cdata = ffi.gc(cparams[0], lib.metkit_free_paramiterator)
+
+        while lib.metkit_paramiterator_next(self._cdata) == lib.METKIT_SUCCESS:
```

Comment:
todo:  next()  should be at the end of this loop, not the beginning, as it skips the first element.
Happens to cancel out an error in the metkit_c, which nullifies the first call to next().